proposed/2022/cpm-dotnet-cli-add-package-support.md

```diff
@@ -4,218 +4,146 @@
-
-> [!NOTE]
-> Scenarios with multiple Directory.Packages.props are out of scope for now.
-
-### 1. The package reference does not exist
-
-If the package does not already exist, it should be added along with the appropriate package version to `Directory.packages.props`. The package version should either be the latest version or the one specified in the CLI command. Only the package name (not the version) should be added to `<PackageReference>` in the project file.
-
-#### Before `add package` is executed
-
-The props file:
-
-```xml
-<Project>
-    <PropertyGroup>
-        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
-    </PropertyGroup>
-    <ItemGroup>
-    </ItemGroup>
-</Project>
-```
-
-The .csproj file:
-
-```xml
-<Project Sdk="Microsoft.NET.Sdk">
-    <PropertyGroup>
-        <TargetFramework>net6.0</TargetFramework>
-        <ImplicitUsings>enable</ImplicitUsings>
-        <Nullable>enable</Nullable>
-    </PropertyGroup>
-
-    <ItemGroup>
-    </ItemGroup>
-</Project>
-```
-
-#### After `add package` is executed
-
-`dotnet add ToDo.csproj package Newtonsoft.Json -v 13.0.1`
-
-The props file:
-
-```xml
-<Project>
-    <PropertyGroup>
-        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
-    </PropertyGroup>
-    <ItemGroup>
-    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1"/>
-    </ItemGroup>
-</Project>
-```
-
-The .csproj file:
-
-```xml
-<Project Sdk="Microsoft.NET.Sdk">
-    <PropertyGroup>
-        <TargetFramework>net6.0</TargetFramework>
-        <ImplicitUsings>enable</ImplicitUsings>
-        <Nullable>enable</Nullable>
-    </PropertyGroup>
-
-    <ItemGroup>
-        <PackageReference Include="Newtonsoft.Json"/>
-    </ItemGroup>
-</Project>
-```
-
-In case there are multiple `Directory.packages.props` files in the repo, the props file that is closest must be considered.
-
-```xml
-Repository
- |-- Directory.Packages.props
- |-- Solution1
+| Scenario # | PackageReference exists? | VersionOverride exists? | PackageVersion exists? | Is Version passed from the commandline? | New behavior in dotnet CLI | In Scope |
+| ---- |----- | ----- | ---- |---- | ----- | ---- |
+| 1 | ❌ | ❌ | ❌ | ❌ | Add `PackageReference` to the project file. Add `PackageVersion` to the `Directory.Packages.Props` file. Use latest version from the package sources. | ✔️ |
+| 2 | ❌ | ❌ | ❌ | ✔️ | Add `PackageReference` to the project file. Add `PackageVersion` to the `Directory.Packages.Props` file. Use version specified in the commandline. | ✔️ |
+| 3 | ❌ | ❌ | ✔️ | ❌ |  Add `PackageReference` to the project file. No changes to the `Directory.Packages.Props` file. Basically we are reusing the version defined centrally for this package. | ✔️ |
+| 4 | ❌ | ❌ | ✔️ | ✔️ | Add `PackageReference` to the project file. Update `PackageVersion` in the `Directory.Packages.Props` file with the version specified in the commandline.  | ✔️ |
```

Comment:
This creates an experience gap which is the `dotnet add package` equivalent for this issue: https://github.com/NuGet/Home/issues/11912. 

@jeffkl Do you want to use the same issue for any potential updates there or should we create a dotnet add package specific one?

Reply:
What I was thinking is that the first 4 scenarios are in scope for @pragnya17's work. The implementation for the first project will be merged into a feature branch https://github.com/NuGet/NuGet.Client/pull/4700. I was wondering if we can address other scenarios and do a single merge into `dev` branch covering all the edge cases listed in the spec. Thoughts?

Reply:
I guess the code path is shared for VS and dotnet.exe. Hence once we address #11912 then it would impact `dotnet add package` also. Please correct if my understanding is incorrect.

Reply:
Not all the code is shared for that actually. 
There's a lot of NuGetPackageManager usage in VS.
VS can also update multiple projects at a time.